### PR TITLE
MAINT: improve error message for isposinf and isneginf on complex values

### PR DESCRIPTION
**Pull request by jor-**
Fixes a bug in `isposinf` and `isneginf` for complex values (#11438) and adds additional tests for these functions with complex values.

### CODE REVIEW

**Comment by pv** [comment]
I'm not sure if these functions should be extended to complex dtypes.

Positive and negative infinity don't really exist in the C99 complex
number model that Numpy follows, in which there's only one complex
infinity.

**Comment by mattip** [comment]
Interesting. We should at least raise an error that makes it clear this is by design. 

@jor- how did you come across this code path, was it part of NumPy, an external library, or just exploring the limits of NumPy's complex handling?

**Comment by jor-** [comment]
Positive and negative infinity can be values in complex NumPy arrays so I think it makes sense that isposinf and isneginf work also for complex arrays.

@mattip: You mean how I found the bug? I use NumPy in one of my libraries to handle complex valued arrays. These arrays might also contain positive and negative infinity which need to be handled separately in my library.

**Comment by eric-wieser** [comment]
> Positive and negative infinity don't really exist in the C99 complex number model

I think this model only makes sense if you provide the `proj` function to normalize the invalid complex infinities, which both C99 and C++11 have:

https://www.ideone.com/IpsC42

Perhaps that has a place in numpy?

**Comment by jor-** [comment]
In my opinion, inf is infinity and inf+nanj is not infinity, so `isposinf` is correct on this. It returns only `True` if the value is `np.inf`.

That `isposinf(a) and isposinf(b)` would not imply `isposinf(a*b)` is due to the specification of the arithmetic and in my option no reason not to implement `isposinf` for complex values.

**Comment by charris** [comment]
I don't think +/- infinity makes sense for complex values.

**Comment by eric-wieser** [comment]
> In my opinion, inf is infinity and inf+nanj is not infinity

Unfortunately, your opinion carries less weight than the C99 standard, which [disagrees with you](https://stackoverflow.com/a/17513954/102441)

> In order to support the one-infinity model, C99 regards any complex value with at least one infinite part as a complex infinity (even if the other part is a NaN)

**Comment by jor-** [comment]
Okay, I did not know about that. Thank you for pointing this out. You have convinced me. With the one infinity model for complex values, the functions `isposinf` and `isneginf` are superfluous or are already covered by `isinf`.

What originally brought me to this problem was, that `+inf` and `-inf` can be stored in complex arrays and thus it looks like the complex plane with `+inf` and `-inf` and not the Riemann sphere with `+inf` only is used.

**Comment by mattip** [comment]
So the conclusion is that calling `np.isposinf` `npisneginf` should raise a `Warning`/`Exception` ?

**Comment by jor-** [comment]
I changed the code so that an exception is raised now if `np.isposinf` or `np.isneginf` is called with complex values.

File: numpy/lib/ufunclike.py [hunk 1]
```diff
@@ -116,8 +116,9 @@ def isposinf(x, out=None):
     NumPy uses the IEEE Standard for Binary Floating-Point for Arithmetic
     (IEEE 754).
 
-    Errors result if the second argument is also supplied when `x` is a
-    scalar input, or if first and second arguments have different shapes.
+    Errors result if the second argument is also supplied when x is a scalar
+    input, if first and second arguments have different shapes, or if the
+    first argument as complex values
```
Comment (mattip) [inline numpy/lib/ufunclike.py]: as -> has, here and in the second docstring

**Comment by mattip** [comment]
Tests that reach the exceptions are missing, you can use `assert_raises(...`

File: numpy/lib/ufunclike.py (hunk 1 above)
Comment (jor-) [reply numpy/lib/ufunclike.py]: changed

**Comment by jor-** [comment]
I added a test. The tests passed. Only one failed due to a failure at Travis.

**Comment by mattip** [comment]
restarted the failing test, now all tests are green.

**Review by mattip — APPROVED** [review]

**Review by eric-wieser — COMMENTED** [review]
Can you justify choosing valueerror instead of typeerror?

File: numpy/lib/ufunclike.py [hunk 2]
```diff
@@ -138,6 +139,10 @@ def isposinf(x, out=None):
     array([0, 0, 1])
 
     """
+    x = nx.asanyarray(x)
```
Comment (mhvk) [inline numpy/lib/ufunclike.py]: As @eric-wieser notes, this leads to a change of behaviour for objects that implement `__array_ufunc__` but cannot be converted to arrays.

**Review by mhvk — CHANGES_REQUESTED** [review]
To avoid the duck-type issue with `np.asanyarray` and also ensure the functions do not become slower for the normal case where the inputs are not complex, I'd suggest using a `try/except`, catching the exception raised by `signbit` and then raising a more informative error message.

**Comment by mhvk** [comment]
p.s. I changed the title of this PR to reflect the consensus that arose that these functions should *not* work on complex values.

p.p.s. I think this is a very nice example of how review makes things better! Thanks for starting it, @jor-!

**Comment by jor-** [comment]
@mhvk: I now use `try/except` instead of `asanyarray`.

@eric-wieser: I changed the error raised to `TypeError` instead of `ValueError`.

**Review by mhvk — APPROVED** [review]

**Comment by mattip** [comment]
Thanks @jor-

**Comment by jor-** [comment]
Thanks for the reviewing.

**Pull request merged**